src/rest/rest_web_server.cpp

```diff
@@ -0,0 +1,233 @@
+    VerifyOrExit(ret >= 0, error = OTBR_ERROR_REST);
+    ret = SetFdNonblocking(mListenFd);
+    VerifyOrExit(ret == 0, error = OTBR_ERROR_REST);
+exit:
+    if (error != OTBR_ERROR_NONE)
+    {
+        otbrLog(OTBR_LOG_ERR, "otbr rest init error");
+    }
+
+    return error;
+}
+
+void RestWebServer::UpdateFdSet(fd_set &aReadFdSet, int &aMaxFd, timeval &aTimeout)
+{
+    // For ReadRdSet
+    UpdateReadFdSet(aReadFdSet, aMaxFd);
+    // For timeout
+    UpdateTimeout(aTimeout, kTimeout, kScaleForCallbackTimeout);
+}
+
+otbrError RestWebServer::Process(fd_set &aReadFdSet)
+{
+    otbrError   error = OTBR_ERROR_NONE;
+    Connection *connection;
+
+    for (auto it = mConnectionSet.begin(); it != mConnectionSet.end(); ++it)
+    {
+        connection = it->second.get();
+
+        if (connection->Readable(kTimeout))
+        {
+            connection->Read();
+        }
+        else if (connection->IsReadTimeout(kTimeout))
+        {
+            connection->Process();
+        }
+        else if (connection->IsCallbackTimeout(kTimeout * kScaleForCallbackTimeout))
+        {
+            connection->CallbackProcess();
+        }
+    }
+
+    error = UpdateConnections(aReadFdSet);
+
+    return error;
+}
+
+void RestWebServer::UpdateTimeout(timeval &aTimeout, int aTimeoutLen, int aScaleForCallback)
+{
+    Connection *   connection;
+    int            timeoutLen;
+    struct timeval timeout = {aTimeout.tv_sec, aTimeout.tv_usec};
+    for (auto it = mConnectionSet.begin(); it != mConnectionSet.end(); ++it)
+    {
+        connection    = it->second.get();
+        timeoutLen    = connection->HasCallback() ? aTimeoutLen * aScaleForCallback : aTimeoutLen;
+        auto duration = duration_cast<microseconds>(steady_clock::now() - connection->GetStartTime()).count();
+        if (duration <= timeoutLen)
+        {
+            timeout.tv_sec = 0;
+
+            if (timeout.tv_usec == 0)
+            {
+                timeout.tv_usec = timeoutLen - duration;
+            }
+            else
+            {
+                timeout.tv_usec = timeout.tv_usec < (timeoutLen - duration) ? timeout.tv_usec : (timeoutLen - duration);
+            }
+        }
+    }
+    if (timercmp(&timeout, &aTimeout, <))
+    {
+        aTimeout = timeout;
+    }
+}
+
+otbrError RestWebServer::UpdateConnections(fd_set &aReadFdSet)
+{
+    otbrError error   = OTBR_ERROR_NONE;
+    socklen_t addrlen = sizeof(sockaddr);
+    int       fd;
+    auto      eraseIt = mConnectionSet.begin();
+    int       ret;
+    for (eraseIt = mConnectionSet.begin(); eraseIt != mConnectionSet.end();)
+    {
+        Connection *connection = eraseIt->second.get();
+
+        if (connection->IsComplete())
+        {
+            eraseIt = mConnectionSet.erase(eraseIt);
+        }
+
+        else
+        {
+            eraseIt++;
+        }
+    }
+
+    if (FD_ISSET(mListenFd, &aReadFdSet))
+    {
+        fd = accept(mListenFd, (struct sockaddr *)mAddress, &addrlen);
```

Comment:
Can `fd` be an invalid value?

Reply:
Besides, write a new method `Accept` or `AcceptNewConnection` to handle the logic.